### DISCUSSION TITLE
Nullable reference types on typeof

### DISCUSSION THREAD

**Discussion by maxtoroq** [dotnet/csharplang#3003] [Language Ideas]
```
var t = typeof(string?);
(1,9): error CS8639: The typeof operator cannot be used on a nullable reference type
```

Why? I don't see any value in throwing this error. Of course `string?` is a type (reference), so why can't I get its type? On the other hand, this breaks code generation scenarios, e.g. `string?` is part of another non-c# file, and I use that snippet to generate `typeof(string?)`.

**Comment by huoyaoyuan** [comment]
Because nullable reference type does not have a standalone runtime representation. When using reflection, it's represented by `NullableContextAttribute` on the target member.

**Comment by CyrusNajmabadi** [comment]
> On the other hand, this breaks code generation scenarios, e.g. string? is part of another non-c# file, and I use that snippet to generate typeof(string?).

Sounds like the code generator should be fixed.

**Comment by maxtoroq** [comment]
> 
> 
> > On the other hand, this breaks code generation scenarios, e.g. string? is part of another non-c# file, and I use that snippet to generate typeof(string?).
> 
> Sounds like the code generator should be fixed.

No shit, what a genius.

**Comment by CyrusNajmabadi** [comment]
In that case, it sounds like the lang/compiler are doing the right thing, and the issue exists outside their purview. 

> Why? I don't see any value in throwing this error. 

This has been answered now. Because supporting code generators is not as valuable as stopping people from writing nonsensical, and likely buggy, code.

**Reply by ManuelKugelmann** [reply]
I think that differentiating between nullable types and non-nullable types should work exactly the same for reference and value types - anything else seems like a design flaw. I think we're getting there, but imho this should be the design goal ...
If you really want to embrace null safe code, nullability has to be a first class citizen of language design. Currently it's a lot better than nothing, but a bit tacked on and brittle in some circumstances.

**Reply by foxesknow** [reply]
I've found it odd how people point out that the `?` goes on the type and doesn't affect the runtime representation, and yet one of the arguments for putting `!!` on the name rather than the type is `!!` [doesn't alter the type system](https://github.com/dotnet/csharplang/discussions/5735#discussioncomment-2142377).

It seems the language team can't make up their minds about these issues. If their stance is that "things that don't alter the type" should go on the parameter then `?` is in the wrong place for nullable reference types.

**Reply by HaloFour** [reply]
@foxesknow 

There's a difference between affecting the type system and affecting the runtime representation.  The compiler does treat `string?` and `string` as different types, but that difference is erased during compilation.  It's like TypeScript types or Java generics in that regard, both of which deeply affect how their respective compilers interpret the types but neither affect the runtime representation or execution of the code.

**Reply by foxesknow** [reply]
> @foxesknow
> 
> There's a difference between affecting the type system and affecting the runtime representation. The compiler does treat `string?` and `string` as different types, but that difference is erased during compilation. It's like TypeScript types or Java generics in that regard, both of which deeply affect how their respective compilers interpret the types but neither affect the runtime representation or execution of the code.

It may affect the compiler, but it doesn't affect the type system. The decision to place `?` on the type, but `!!` on the parameter is inconsistent. Neither affects the type system.

**Reply by HaloFour** [reply]
@foxesknow 

> It may affect the compiler, but it doesn't affect the type system. 

It affects the type system as far as the compiler is concerned.  The fact that it doesn't affect the type system of the runtime doesn't really matter.  The same is true of `dynamic`, which doesn't exist in the runtime and is compiled down to plain-old `object`, but that also has a dramatic effect on the type system.

**Reply by jnm2** [reply]
@foxesknow The decision to place `!!` on the parameter name is the most consistent. If it could be placed on a type, it would imply that it would make sense in other contexts where there are types, including things like `List<string!!>`. The parameter is the actual thing being targeted by it.

It's also most consistent with potential future plans to have things like `nullableInt!!` instead of `nullableInt!.Value`. It's not about the type; it's about the variable being checked.

**Reply by foxesknow** [reply]
> @foxesknow The decision to place `!!` on the parameter name is the most consistent. If it could be placed on a type, it would imply that it would make sense in other contexts where there are types, including things like `List<string!!>`. The parameter is the actual thing being targeted by it.
> 
> It's also most consistent with potential future plans to have things like `nullableInt!!` instead of `nullableInt!.Value`. It's not about the type; it's about the variable being checked.


Your example with `List<string!!>` actually highlights an underlying problem, which is you can't create a `List<>` where you can be sure that it doesn't contains a `null`. In a nullable context it's all to easy for a `null` to end up being passed through for a non-nullable reference type. It's crazy than in a `#nullable enabled` context you still have to check that `Add(string s)` isn't `null`.

The band-aid approach to adding language features creates a mess.

**Reply by HaloFour** [reply]
@foxesknow 

> Your example with `List<string!!>` actually highlights an underlying problem, which is you can't create a `List<>` where you can be sure that it doesn't contains a null.

Because that is not and will never be possible in .NET.  That's the reality of working with a 20 year old runtime, one where `null` is often the default value, and where you might have to interop with any number of a dozen different languages that have no concept of C# language features.

**Reply by CyrusNajmabadi** [reply]
> Your example with List<string!!> actually highlights an underlying problem, which is you can't create a List<> where you can be sure that it doesn't contains a null

Correct. You cannot.  That's exactly why we did not put it on the type. Because there's no way to actually get those semantics with .net. 

Note that List is a prime example of why this cannot work. A List is backed by an array with more elements than what the List actually contains.  If the List only contained non-null values this would be impossible to represent as you'd have an illegal array that said it didn't have null entries, but would need exactly that for the  extra space in it currently not filled with any actual user supplied values.

**Reply by CyrusNajmabadi** [reply]
> The band-aid approach to adding language features creates a mess.

Yes. Welcome to the difficulty of dealing with billions of lines of code that depends on these semantics from the last 20 years. We cannot break compat on all that. So that means accepting what the real semantics of .net are and having tools to assist that do not come with a compat burden that would make future versions completely untenable for large parts of the ecosystem. 

This can be disappointing to those who wound be willing to accept hard breaks. But that group is a tiny minority compared to the overall community.

**Reply by foxesknow** [reply]
> > Your example with List<string!!> actually highlights an underlying problem, which is you can't create a List<> where you can be sure that it doesn't contains a null
> 
> Correct. You cannot. That's exactly why we did not put it on the type. Because there's no way to actually get those semantics with .net.
> 
> Note that List is a prime example of why this cannot work. A List is backed by an array with more elements than what the List actually contains. If the List only contained non-null values this would be impossible to represent as you'd have an illegal array that said it didn't have null entries, but would need exactly that for the extra space in it currently not filled with any actual user supplied values.

That's an interesting point you make about the underlying array that backs a `List<>`. However, in a nullable context a `List<string>` is also backed by an array that can contain null values in the extra space that isn't filled with user supplied values.

**Reply by CyrusNajmabadi** [reply]
Yes.  But there's no way to represent that.  Consider how you might actually state that in a type signature.  If i have `List<T>`  i cannot back it with a `T[]` as this wouldn't work if i ended up instantiating a `List<T!!>` (presuming such a feature existed).  So i'd have to back it with a `T?[]`.  However, then we're in the exact scenario you were unhappy with where: `It's crazy than in a #nullable enabled context you still have to check that ... isn't null`. 

E.g. because you'd have to state that the underlying array could have null elements, your code would still absolutely need null checks to make sure you weren't violating the contract of the code that instantiated you.  The compiler could not validate this because it woudl have no way to know that some values in the array were legal to index and give you non-null back, while some indices absolutely could give null and would have to be checked.  

This is why non-null cannot currently be part of the real runtime representation of things.  Because there is genuinely no way to actually get this in a system like .net.

**Comment by richlander** [comment]
@maxtoroq -- I hid your comment as "Abuse". It was reported as a [conduct of conduct](https://dotnetfoundation.org/code-of-conduct) violation and is abusive. Please use more constructive language.
